### DISCUSSION TITLE
Recursive toArray

### DISCUSSION THREAD

**Discussion by godismyjudge95** [spatie/laravel-collection-macros#237] [Ideas]
I don't know if this is doable but I found a need for doing a recursive toArray on a collection (nested objects that are able to be converted to arrays).

If anyone has an idea of how to do this with existing methods that would be awesome.

**Comment by the-pulli** [comment]
I'm not sure, if that's what you mean. I created my own macro called `mapToCollection` which recursively wraps nested arrays and objects with a toArray method to a Collection object, which allows nested access.

You can find the repo here: https://github.com/the-pulli/collection-macros

Maybe the tests give you a clue, what I mean: https://github.com/the-pulli/collection-macros/blob/main/tests/Macros/MapToCollectionTest.php

Because I mainly use Collections outside of Laravel, I haven't created a PR (see: https://github.com/spatie/laravel-collection-macros/discussions/246)

But I would be happy to do so, if there is a need for this kind of macro.

**Reply by godismyjudge95** [reply]
Not sure that is what I am looking for.  My need is for something like this:

```php
$items = collect([
	new Parent(
		name: 'parent1',
		children: collect([new Child(name: 'child1')]),
		other: collect([new Other(value: 'other1')]),
	),
	new Parent(
		name: 'parent2',
		children: collect([new Child(name: 'child2')]),
		other: collect([new Other(value: 'other2')]),
	),
]);
```

to turn all of those classes and subclasses to arrays:
```php
return $items->recursiveToArray();
```

Which would output:
```php
[
	[
		'name' => 'parent1',
		'children' => [
			['name' => 'child1'],
		],
		'other' => [
			['value' => 'other1'],
		],
	],
	[
		'name' => 'parent2',
		'children' => [
			['name' => 'child2'],
		],
		'other' => [
			['value' => 'other2'],
		],
	],
]
```

**Reply by the-pulli** [reply]
I just created a test case with my macro for this scenario. It's working, but you have to implement the toArray method individually on each class.

See it in action in the corresponding test: 

https://github.com/the-pulli/collection-macros/blob/8c9d2e66f18e5a52770d8909cc89c522742abb25/tests/Macros/MapToCollectionTest.php#L62

You then call it like this:

```php
$data = Collection::mapToCollection([your data here], true);
// because you get a collection object back, you can do whatever you want with it
// if you just need the data as array:
$data->toArray();
```

**Reply by the-pulli** [reply]
I just added your recursiveToArray macro to my repo.

Now you can just call on the collection:

```php
Collection::recursiveToArray([your data here]);
```

and under the hood it calls mapToCollection and converts it directly to an array.

See the test here:

https://github.com/the-pulli/collection-macros/blob/main/tests/Macros/RecursiveToArrayTest.php
